File: 1405-longest-happy-string/1405-longest-happy-string.cs
public class Solution {
    public string LongestDiverseString(int a, int b, int c) {
        
            var pq = new PriorityQueue<string, int>(Comparer<int>.Create((a, b) => b.CompareTo(a)));
            pq.Enqueue("a", a);
            pq.Enqueue("b", b);
            pq.Enqueue("c", c);
            StringBuilder sb = new StringBuilder();
            string prev = "";
            int prevc = 0;

            while (pq.Count > 0)
            {
                if(pq.TryDequeue(out string ele,out int count) && count > 0)
                {
                    int max = 0;
                    if (count >= prevc)
                    {
                        while (count != 0 && max != 2)
                        {
                            sb.Append(ele);
                            count--;
                            max++;
                        }
                    }
                    else
                    {
                        sb.Append(ele);
                        count--;
                    }
                    if (prevc > 0)
                    {
                        pq.Enqueue(prev, prevc);
                    }
                    prev = ele;
                    prevc = count;
                }
            }
            return sb.ToString();
    }
}

File: 1405-longest-happy-string/1405-longest-happy-string.cpp
class Solution {
public:
    string longestDiverseString(int a, int b, int c) {
        priority_queue<pair<int,char>> q;
        if(a) q.push({a,'a'});
        if(b) q.push({b,'b'});
        if(c) q.push({c,'c'});
        string res="";
        while (!q.empty())
        {
            pair<int,char> p = q.top();
            q.pop();
            int c=p.first;
            char ch=p.second;
            if(res.size()>=2 && res[res.size()-1]==ch && res[res.size()-2]==ch){
                if(q.empty()) break;
                pair<int,char> a = q.top();
                q.pop();
                res+=a.second;
                if(a.first-1>0){
                    q.push({a.first-1,a.second});
                }
            }else{
                c--;
                res+=ch;
            }
            if(c>0){
                q.push({c,ch});
            }
        }        
        return res;
    }
};